### PR TITLE
feat: support granting explicit permissions of a resource to multiple users in a transaction (AM-218)

### PR DESCRIPTION
**Pull request by rebeccabrennan**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/AM-218
(Being updated to follow agreed structure)

### CODE REVIEW

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java [hunk 1]
```diff
@@ -6,14 +6,15 @@
...
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
 @Data
 @Builder
 public final class ExplicitAccessGrant {
     private final String resourceId;
-    private final String accessorId;
+    private final List<String> accessorId;
```
Comment (geek-soft) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java]: Please rename since you turned singular field into list .

Comment (geek-soft) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java]: Shouldn't that be a set?

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java [hunk 2]
```diff
@@ -69,22 +69,25 @@ public void grantExplicitResourceAccess(ExplicitAccessGrant explicitAccessGrant)
             throw new IllegalArgumentException("At least one permission per attribute is required");
         }
 
+
```
Comment (geek-soft) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: This again?

File: am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java [hunk 3]
```diff
@@ -27,17 +29,17 @@ private TestDataFactory() {
 
     public static ExplicitAccessGrant createGrantForWholeDocument(String resourceId,
                                                                   Set<Permission> permissions) {
-        return createGrantForWholeDocument(resourceId, ACCESSOR_ID, permissions);
+        return createGrantForWholeDocument(resourceId, ACCESSOR_IDS, permissions);
     }
 
     public static ExplicitAccessGrant createGrantForWholeDocument(String resourceId,
-                                                                  String accessorId,
+                                                                  List<String> accessorId,
```
Comment (geek-soft) [inline am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java]: Why to change it? Can't we have overloaded method that accepts single user ID?

Comment (rebeccabrennan) [reply am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java]: 👍

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java (hunk 2 above)
Comment (rebeccabrennan) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: 👍

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java (hunk 1 above)
Comment (rebeccabrennan) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java]: 👍

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java [hunk 4]
```diff
@@ -69,22 +71,29 @@ public void grantExplicitResourceAccess(ExplicitAccessGrant explicitAccessGrant)
             throw new IllegalArgumentException("At least one permission per attribute is required");
         }
 
+        if (explicitAccessGrant.getAccessorIds().stream()
```
Comment (geek-soft) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: Throw away code :P I am implementing validations in a different way.

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java [hunk 5]
```diff
@@ -127,7 +136,7 @@ public void revokeResourceAccess(ExplicitAccessMetadata explicitAccessMetadata)
      * @param userRoles accessor roles
      * @param resources envelope {@link Resource} and corresponding metadata
      * @return envelope list of {@link FilterResourceResponse} with resource ID, filtered JSON and map of permissions
-     *     if access to resource is configured, otherwise null.
+     *      if access to resource is configured, otherwise null.
```
Comment (geek-soft) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: Why is that change there? Please try to avoid unnecessary changes that does nothing. It just takes time from reviewer.

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java [hunk 6]
```diff
@@ -13,7 +13,7 @@
 @Builder
 public final class ExplicitAccessGrant {
     private final String resourceId;
-    private final String accessorId;
+    private final Set<String> accessorIds;
```
Comment (geek-soft) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java]: In my mind validation will be done in the following way (just FYI)

```
private final @NotEmpty Set<@NotBlank String> accessorIds;
```

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/FilterResourceIntegrationTest.java [hunk 7]
```diff
@@ -42,7 +43,7 @@ void setUp() {
 
     @Test
     void whenRowExistsAndHasReadPermissionsShouldReturnEnvelopeWithData() {
-        service.grantExplicitResourceAccess(createGrantForWholeDocument(resourceId, ACCESSOR_ID, READ_PERMISSION));
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/FilterResourceIntegrationTest.java]: I suggest reverting these changes. There is really no point changing all methods to accept set now. Let's take one element by default and lets overload method to accept list if needed. For most test cases it does not matter whether there is one or more accessors. Unnecessary changes IMO.

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java (hunk 5 above)
Comment (rebeccabrennan) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: checkstyle wanted it moved.

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java (hunk 6 above)
Comment (rebeccabrennan) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java]: 👍

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/FilterResourceIntegrationTest.java (hunk 7 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/FilterResourceIntegrationTest.java]: 👍

Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/FilterResourceIntegrationTest.java]: Changes were made before overloaded :) can change back now.

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java [hunk 8]
```diff
@@ -26,26 +28,28 @@ void setUp() {
     @Test
     @SuppressWarnings("LineLength")
     void ifUserHasAccessShouldReturnUserIds() {
-        service.grantExplicitResourceAccess(createGrantForWholeDocument(resourceId, ACCESSOR_ID, READ_PERMISSION));
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java]: How many times I need to ask to get it reverted?

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java [hunk 9]
```diff
@@ -75,4 +77,29 @@ void whenCreatingDuplicateResourceAccessEntryIsOverwritten() {
 
         assertThat(databaseHelper.countExplicitPermissions(resourceId)).isEqualTo(1);
     }
+
+    @Test
+    void whenGrantingAccessForMultipleUsersEntriesShouldAppearInDatabase() {
+        Map<JsonPointer, Set<Permission>> multipleAttributePermissions = new ConcurrentHashMap<>();
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: No more ConcurrentHashMap please. We have Guava in test scope.

Comment (geek-soft) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: Why not to use one of the methods that grant access to whole doc? Does it matter for this particular test what attribute access is granted for?

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java [hunk 10]
```diff
@@ -75,4 +77,29 @@ void whenCreatingDuplicateResourceAccessEntryIsOverwritten() {
...
+
+        service.grantExplicitResourceAccess(createGrant(resourceId, MULTIPLE_ACCESSOR_IDS,
+            multipleAttributePermissions));
+
+        assertThat(databaseHelper.findExplicitPermissions(resourceId)).hasSize(3)
+            .extracting(ExplicitAccessRecord::getAccessorId).contains("a","b","c");
+    }
+
+    @Test
+    void whenGrantingAccessForMultipleUsersAndMultipleAttributesEntriesShouldAppearInDatabase() {
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: What value does the test have on top of the previous one? Does it cover new execution path?

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java (hunk 9 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: 👍

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java (hunk 10 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: Took out before this comment made :D

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java (hunk 8 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java]: 👍

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java (hunk 4 above)
Comment (rebeccabrennan) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: 👍

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java [hunk 11]
```diff
@@ -17,8 +17,8 @@
 public final class ExplicitAccessGrant {
     @NotBlank
     private final String resourceId;
-    @NotBlank
-    private final String accessorId;
+    @NotEmpty
+    private final Set<@NotNull String> accessorIds;
```
Comment (geek-soft) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java]: Use `@Not blank` on set items.

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java [hunk 12]
```diff
@@ -27,7 +28,8 @@ void setUp() {
     @SuppressWarnings("LineLength")
     void ifUserHasAccessShouldReturnUserIds() {
         service.grantExplicitResourceAccess(createGrantForWholeDocument(resourceId, ACCESSOR_ID, READ_PERMISSION));
-        service.grantExplicitResourceAccess(createGrantForWholeDocument(resourceId, OTHER_ACCESSOR_ID, READ_PERMISSION));
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java]: Revert

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java [hunk 13]
```diff
@@ -36,7 +38,7 @@ void ifUserHasAccessShouldReturnUserIds() {
 
     @Test
     void ifUserHasNoAccessShouldReturnNull() {
-        service.grantExplicitResourceAccess(createGrantForWholeDocument(resourceId, ACCESSOR_ID, READ_PERMISSION));
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java]: Revert

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java [hunk 14]
```diff
@@ -45,7 +47,7 @@ void ifUserHasNoAccessShouldReturnNull() {
 
     @Test
     void whenCheckingAccessToNonExistingResourceShouldReturnNull() {
-        service.grantExplicitResourceAccess(createGrantForWholeDocument(resourceId, ACCESSOR_ID, READ_PERMISSION));
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java]: Revert

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java (hunk 11 above)
Comment (rebeccabrennan) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java]: 👍

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java (hunk 14 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java]: 👍

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java (hunk 13 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java]: 👍

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java (hunk 12 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java]: 👍

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java (hunk 11 above)
Comment (rebeccabrennan) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ExplicitAccessGrant.java]: 👍

File: am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java [hunk 15]
```diff
@@ -42,7 +52,22 @@ public static ExplicitAccessGrant createGrant(String resourceId,
                                                   Map<JsonPointer, Set<Permission>> attributePermissions) {
         return ExplicitAccessGrant.builder()
```
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java]: 👍

File: am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java [hunk 16]
```diff
@@ -72,7 +97,7 @@ public static ExplicitAccessMetadata createMetadata(String resourceId) {
             .resourceType(RESOURCE_TYPE)
             .resourceName(RESOURCE_NAME)
             .attribute(JsonPointer.valueOf(""))
-            .securityClassification(SecurityClassification.PUBLIC)
+            .securityClassification(SecurityClassification.PRIVATE)
```
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java]: Changed when trying to figure out security classification was coming back as null. Fixed now though.

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java (hunk 12 above)
Comment (geek-soft) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GetAccessorsListIntegrationTest.java]: I asked for revert, not to remove that call. I simply do not want to see changes in this class as they are irrelevant.

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java [hunk 17]
```diff
@@ -37,13 +38,16 @@ void whenCreatingResourceAccessResourceAccessAppearsInDatabase() {
 
     @Test
     void whenCreatingResourceAccessMultipleEntriesAppearInDatabase() {
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: In `GrantAccessIntegrationTest.java` there should be test showing that it is possible to grant access to multiple users. One test only.

File: am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestConstants.java [hunk 18]
```diff
@@ -25,6 +25,7 @@
     public static final JsonPointer ATTRIBUTE = JsonPointer.valueOf("/test");
     public static final String ACCESSOR_ID = "a";
     public static final String OTHER_ACCESSOR_ID = "b";
+    public static final Set<String> ACCESSOR_IDS = Stream.of("y","z").collect(toSet());
```
Comment (geek-soft) [inline am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestConstants.java]: Space between items please

Comment (geek-soft) [reply am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestConstants.java]: Use ImmutableSet#of as well.

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java [hunk 19]
```diff
@@ -53,4 +56,18 @@ void whenCreatingDuplicateResourceAccessEntryIsOverwritten() {
...
+        Map<JsonPointer, Set<Permission>> multipleAttributePermissions =
+            ImmutableMap.<JsonPointer, Set<Permission>>builder()
+                .put(JsonPointer.valueOf(""), EXPLICIT_READ_CREATE_UPDATE_PERMISSIONS)
+                .put(JsonPointer.valueOf("/name"), EXPLICIT_READ_CREATE_UPDATE_PERMISSIONS)
+                .build();
+
+        service.grantExplicitResourceAccess(createGrant(resourceId, ACCESSOR_IDS, multipleAttributePermissions));
+
+        assertThat(databaseHelper.findExplicitPermissions(resourceId)).hasSize(4)
+            .extracting(ExplicitAccessRecord::getAccessorId).contains("y", "z");
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: Be more string. We wanna make sure only Y and Z are there.

File: am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java [hunk 20]
```diff
@@ -34,15 +37,22 @@ public static ExplicitAccessGrant createGrantForWholeDocument(String resourceId,
     public static ExplicitAccessGrant createGrantForWholeDocument(String resourceId,
                                                                   String accessorId,
                                                                   Set<Permission> permissions) {
+        return createGrant(resourceId, Stream.of(accessorId).collect(Collectors.toSet()),
```
Comment (geek-soft) [inline am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java]: No `Stream.of` please. Use guava lib.

Comment (geek-soft) [reply am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java]: Delegate to `createGrantForWholeDocument(String resourceId, Set<String> accessorId, Set<Permission> permission)`

File: am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java [hunk 21]
```diff
@@ -52,6 +62,12 @@ public static ExplicitAccessGrant createGrant(String resourceId,
             .build();
     }
 
+    public static ExplicitAccessGrant createGrant(String resourceId,
```
Comment (geek-soft) [inline am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java]: Move above `createGrant(String resourceId, Set<String> accessorId, Map<JsonPointer, Set<Permission>> attributePermissions)` so it reads more naturally.

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java (hunk 19 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: 👍

File: am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java (hunk 21 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java]: 👍

File: am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java (hunk 20 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestDataFactory.java]: 👍

File: am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestConstants.java (hunk 18 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/uk/gov/hmcts/reform/amlib/helpers/TestConstants.java]: 👍

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java (hunk 17 above)
Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: 👍

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java (hunk 5 above)
Comment (geek-soft) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: Checkstyle expects 4 whitespaces while you got 5. Please revert as previous state should satisfy style check.

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java [hunk 22]
```diff
@@ -53,4 +56,17 @@ void whenCreatingDuplicateResourceAccessEntryIsOverwritten() {
 
         assertThat(databaseHelper.countExplicitPermissions(resourceId)).isEqualTo(1);
     }
+
+    @Test
+    void whenCreatingResourceForMultipleUsersShouldAppearInDatabase() {
+        Map<JsonPointer, Set<Permission>> multipleAttributePermissions =
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: That name does not reflect real state. There are no multiple attributes in array, just one.

Why not to use `createGrantForWholeDocument` in this test? Does attribute matter for that test case? I do not think so.

Comment (rebeccabrennan) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/GrantAccessIntegrationTest.java]: 👍

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java (hunk 5 above)
Comment (rebeccabrennan) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: 👍
